### PR TITLE
Prediction Explanations: Force Plots

### PR DESCRIPTION
**Pull request by chukarsten**
Addresses #1970 

Single row explanations force plots:

Screen Shot 2021-04-21 at 1 19 18 PM

### CODE REVIEW

File: evalml/model_understanding/graphs.py
Comment (freddyaboulton) [inline evalml/model_understanding/graphs.py:1178]: The issue mentions adding a "data method" to return the data needed to create the plot in json plot. Is that planned for a separate PR?

Comment (chukarsten) [reply evalml/model_understanding/graphs.py:1178]: I need to chat with @dsherry about what exactly that entails...I think I'll need some specifics but I planned to do that here.

Comment (dsherry) [reply evalml/model_understanding/graphs.py:1178]: Our model understanding tools should be broken into two chunks:
* Data generation: given necessary inputs, return all the information required to visualize the results
* Visualization: call the data generation method, and use pyplot or similar to display the results

Separating these two steps is important because it means that the data generation method could be used to display the same information in a different UI.

File: evalml/model_understanding/prediction_explanations/_user_interface.py
Comment (chukarsten) [inline evalml/model_understanding/prediction_explanations/_user_interface.py:193]: I came across an instance of the data being passed through as a list, like it seems to expect.

File: evalml/model_understanding/prediction_explanations/_user_interface.py
Comment (chukarsten) [inline evalml/model_understanding/prediction_explanations/_user_interface.py:196]: In the similar style as drill down.

File: evalml/model_understanding/prediction_explanations/_user_interface.py
Comment (freddyaboulton) [inline evalml/model_understanding/prediction_explanations/_user_interface.py:14]: `include_expected_values` is not used. I think we can get rid of it?

File: evalml/model_understanding/prediction_explanations/explainers.py
Comment (freddyaboulton) [inline evalml/model_understanding/prediction_explanations/explainers.py:73]: I think there is a bug here because the expected value is included regardless of the value of this parameter

```python
exp = explain_predictions(pipeline=pipeline, input_features=X, y=None, indices_to_explain=[3],
                           top_k_features=6, include_shap_values=True, output_format="dict",
                   include_expected_value=False)
for explanation in exp["explanations"]:
    for class_explanation in explanation["explanations"]:
        assert class_explanation["expected_value"] != 0
```

I think in this case it's fine to not surface this parameter to the top-level and instead always include the expected value in the dict output. The expected value really only makes sense for the dict output format so I worry that including this parameter will cause users who use the dataframe or text output to be confused.

**Review by freddyaboulton — APPROVED** [review]
@chukarsten Thank you! I think it's great we can use `explain_predictions` as our data function.

I left a couple of questions/suggestions I'd like to resolve before merging.

File: evalml/model_understanding/force_plots.py
Comment (chukarsten) [reply evalml/model_understanding/force_plots.py:63]: Yup!

File: docs/source/release_notes.rst
Comment (dsherry) [reply docs/source/release_notes.rst:83]: Please also mention "prediction explanations"
> Added force plots for prediction explanations

File: docs/source/user_guide/model_understanding.ipynb
Comment (dsherry) [inline docs/source/user_guide/model_understanding.ipynb:593]: You don't need to fit a pipeline here. We fit one at the top of the model understanding guide. Please delete.

https://evalml.alteryx.com/en/v0.24.1/user_guide/model_understanding.html

File: docs/source/user_guide/model_understanding.ipynb
Comment (dsherry) [inline docs/source/user_guide/model_understanding.ipynb:595]: Is there a reason why our code shouldn't automate this?

We have a method called `jupyter_check` which should tell you whether or not code is being run in an ipython environment. If true, could `graph_force_plot` call `shap.initjs()` internally?

Relatedly, so far we've only used `jupyter_check` to conditionally import `ipywidgets`, which I believe we do for similar reasons to this call to `shap.initjs()`... would doing that in `graph_force_plot` make this simpler?

Here's an example: https://github.com/alteryx/evalml/blob/main/evalml/model_understanding/graphs.py#L909

Also, I took a look at [what `initjs` is doing](https://github.com/slundberg/shap/blob/master/shap/plots/_force.py#L244). It looks like it checks that ipython is set up, imports the shap logo and displays the logo in an HTML div alongside a javascript bundle.

File: docs/source/user_guide/model_understanding.ipynb
Comment (dsherry) [inline docs/source/user_guide/model_understanding.ipynb:626]: Style nit-pick: if its important to show a multiclass example for force plots, I suggest you:
* Rename the variable name of the binary classification pipeline which all the rest of the model understanding tools use to `pipeline_binary`
* At the top of the notebook where we define the binary pipeline, move this code up there so we also define `pipeline_multiclass`
* Then below you can use `pipeline_multiclass`

The reason I like this is that it keeps the notebook's narrative super simple. First train models, then visualize.

File: evalml/model_understanding/force_plots.py
Comment (dsherry) [reply evalml/model_understanding/force_plots.py:9]: +1, we should follow the convention our other graph methods follow unless there's a reason to do otherwise.

I think returning a plot is super useful for jupyter notebooks actually -- they'll auto-render the last returned value.

File: evalml/model_understanding/force_plots.py
Comment (dsherry) [inline evalml/model_understanding/force_plots.py:31]: Do we have an issue filed for this TODO?

File: evalml/model_understanding/force_plots.py
Comment (dsherry) [reply evalml/model_understanding/force_plots.py:9]: Ah, I read a bit further. I see this is because both `graph_force_plot` and `force_plot` accept a list of rows to explain rather than a single row.

I suggest we define `force_plot` and `graph_force_plot` to accept a single row. Then define `force_plots` and `graph_force_plots` to accept multiple and to call the single-row methods.

File: evalml/model_understanding/force_plots.py
Comment (dsherry) [inline evalml/model_understanding/force_plots.py:72]: Is the value of this `'plot'` key a class or a string? I ask because if its a class, we can't easily JSON-serialize this output, so we'd likely have to convert to string somewhere in order for a web UI to consume this.

File: evalml/model_understanding/force_plots.py
Comment (dsherry) [inline evalml/model_understanding/force_plots.py:93]: Simplification: use the [dict `get` method](https://docs.python.org/3/library/stdtypes.html#dict.get)
```python
cls = cls_exp.get("class_name", "regression")
```

Comment (dsherry) [inline evalml/model_understanding/force_plots.py:93]: So, according to this logic, if the `"class_name"` comes back as `None` here, and we set `cls` to `"regression"`, doesn't that mean the assignment to `row_exp_dict[cls]` will be overwriting the same key for every loop in this `for` loop?

File: evalml/model_understanding/prediction_explanations/_algorithms.py
Comment (dsherry) [inline evalml/model_understanding/prediction_explanations/_algorithms.py:106]: Returning variable length arguments can make consuming output a bit uglier up the stack. I suggest we always return a tuple of two arguments. If `explainer` here happens to return as `None`, that's fine.

Rolling with @freddyaboulton 's suggestion above though, I wonder if we can just delete `return_explainer` and simply always return both the dictionary and the explainer here.

File: evalml/model_understanding/prediction_explanations/_algorithms.py
Comment (dsherry) [reply evalml/model_understanding/prediction_explanations/_algorithms.py:34]: +1

Why add this flag? If it costs us extra CPU cycles to compute the explainer to be returned, I think that's a good argument for having a flag. But at first glance through this code, I don't think that's the case.

**Review by dsherry — CHANGES_REQUESTED** [review]
Looking solid, left a few comments

File: evalml/model_understanding/prediction_explanations/_algorithms.py
Comment (chukarsten) [reply evalml/model_understanding/prediction_explanations/_algorithms.py:34]: I don't think so, it gets called in a lot of other tests when the expected value is not desired (non-force plots).

File: docs/source/release_notes.rst
Comment (chukarsten) [reply docs/source/release_notes.rst:83]: I'm legit mystified how I put this in the wrong section.

File: docs/source/user_guide/model_understanding.ipynb
Comment (chukarsten) [reply docs/source/user_guide/model_understanding.ipynb:626]: gucci

File: docs/source/user_guide/model_understanding.ipynb
Comment (chukarsten) [reply docs/source/user_guide/model_understanding.ipynb]: Done and done!

File: evalml/model_understanding/prediction_explanations/_algorithms.py
Comment (chukarsten) [reply evalml/model_understanding/prediction_explanations/_algorithms.py]: Done, this is now returning the expected_value instead of the explainer.  Always.

File: evalml/model_understanding/force_plots.py
Comment (chukarsten) [reply evalml/model_understanding/force_plots.py:32]: @dsherry yup, we traiged this : https://github.com/alteryx/evalml/issues/2287

File: evalml/model_understanding/force_plots.py
Comment (chukarsten) [reply evalml/model_understanding/force_plots.py:77]: I didn't think the Web UI was going to consume this.  My understanding was that the graphing function was going to be for the purpose of EvalML users via jupyter notebook, etc. and that the data function, which didn't add the plot to the dict was going to be the one driving the web UI (which was going to render the force plots itself to get all the nice javascript stuff that shap does)

File: evalml/model_understanding/force_plots.py
Comment (chukarsten) [reply evalml/model_understanding/force_plots.py:98]: The possible values for the key "class_name" are None, the name of the positive class ("malignant" in these tests), or the names of the different classes ["class_0", "class_1", "class_2"].  the row_exp variable is a list that contains either 1 explanation for regression type problems, 1 explanation for binary class problems (the positive class) or multiple explanations for multi-class problems.  If the row for the regression type problems had more than one "class" to explain then yes, it would continuously overwrite that key, but in this case it should be good.

File: docs/source/user_guide/model_understanding.ipynb
Comment (chukarsten) [reply docs/source/user_guide/model_understanding.ipynb]: I made the changes to the pipeline names, but some of the visualization, particularly with the binary pipelines, is spread out all over the notebook.  I renamed them, but kept them in the same spot for now.

File: evalml/model_understanding/force_plots.py
Comment (chukarsten) [reply evalml/model_understanding/force_plots.py:98]: Haha you got me on this one!  "class_name" always exists, but it's the *value* of the "class_name" key that is None.  The .get() function is for if the "class_name" key doesn't exist.  You taught me something here :)

File: evalml/model_understanding/force_plots.py
Comment (chukarsten) [reply evalml/model_understanding/force_plots.py:10]: @dsherry I think I'm going to let #2287 handle the multi row case.  shap._force gives us back a different style of visualizer for multi-row explanations than it does for single row explanations.  How I handle that now is to provide the single row visualization for each row rather than the combined multi-row explanation.  Originally, before I used explain_predictions, `graph_force_plots` properly returned the visualization for both cases, making separate graphing functions unnecessary.  I hope after #2287 is complete, it will again! 

However, I will change the return to just be the plots rather than all the other junk.

Comment (chukarsten) [reply evalml/model_understanding/force_plots.py:10]: @freddyaboulton I am not entirely sure what this AdditiveForceVisualizer is.  It seems like it's a python wrapper for some javascript visualization.  Which makes me unsure if I can wrap it into a plotly figure.

File: evalml/model_understanding/prediction_explanations/_user_interface.py
Comment (freddyaboulton) [inline evalml/model_understanding/prediction_explanations/_user_interface.py:732]: I don't think the value of this parameter is actually used in the implementation. I vote we remove it from the function signature.

File: evalml/model_understanding/prediction_explanations/_user_interface.py
Comment (freddyaboulton) [inline evalml/model_understanding/prediction_explanations/_user_interface.py:264]: nit-pick: As things are implemented now, the `expected_value` is always a key so we can get rid of the check.

File: evalml/model_understanding/prediction_explanations/_user_interface.py
Comment (freddyaboulton) [inline evalml/model_understanding/prediction_explanations/_user_interface.py:128]: I think we can get rid of this parameter since it's not used.

File: evalml/tests/model_understanding_tests/prediction_explanations_tests/test_force_plots.py
Comment (freddyaboulton) [inline evalml/tests/model_understanding_tests/prediction_explanations_tests/test_force_plots.py:28]: 😂

File: evalml/tests/model_understanding_tests/prediction_explanations_tests/test_force_plots.py
Comment (freddyaboulton) [inline evalml/tests/model_understanding_tests/prediction_explanations_tests/test_force_plots.py:39]: Do you think it's worth adding a test to make sure the force plot displays the aggregate contribution of features that have been OHE/text features? I vote yes just because I think it's important to make sure we maintain that behavior since the reports also aggregate the OHE/text features.

**Review by freddyaboulton — APPROVED** [review]
@chukarsten Thank you for making the changes! I think one clean-up I'd like to do is remove the `_include_expected_value` parameter since it's not used by the forceplot code and it doesn't seem to change the output of the explainers.

I left a comment about adding some tests for pipelines that use OHE/Text Featurizers so that we check the features are aggregated. Let me know what you think about this change.

I also think there may be a bug with the jupyter display? The following code raises `TypeError: Object of type Timestamp is not JSON serializable`:

```python
from evalml.demos import load_fraud
from evalml.pipelines import BinaryClassificationPipeline
from evalml.model_understanding import explain_predictions
from evalml.model_understanding.force_plots import graph_force_plot
import pytest


X, y = load_fraud(1000)

pipeline = BinaryClassificationPipeline(["Text Featurization Component", "DateTime Featurization Component",
                                         "One Hot Encoder", "Random Forest Classifier"])
pipeline.fit(X, y)

results = graph_force_plot(pipeline, rows_to_explain=[1],
                           training_data=X, y=y)

for result in results:
    for cls in result:
        print("Class:", cls)
            display(result[cls]["plot"])
```

image

I think the issue is that the input features are timestamps: 

image


I say we don't block merge on that bug though cause it's coming from the `display` call. My hypothesis is that it might work if users use their own code to display the data from `force_plot`. I think we should file a spike issue to investigate further. Let me know what you think.

File: docs/source/release_notes.rst
Comment (chukarsten) [reply docs/source/release_notes.rst:41]: this has lingered so long

File: evalml/model_understanding/prediction_explanations/_user_interface.py
Comment (chukarsten) [reply evalml/model_understanding/prediction_explanations/_user_interface.py:128]: not sure how that got past....

File: evalml/tests/model_understanding_tests/prediction_explanations_tests/test_force_plots.py
Comment (chukarsten) [reply evalml/tests/model_understanding_tests/prediction_explanations_tests/test_force_plots.py:39]: I do....I just need to figure out what that looks like, haha.

File: evalml/model_understanding/force_plots.py
Comment (dsherry) [reply evalml/model_understanding/force_plots.py:84]: Agreed. But isn't `force_plot` (aka this code) the data function and `graph_force_plot` the graphing function?

File: evalml/model_understanding/prediction_explanations/_algorithms.py
Comment (dsherry) [inline evalml/model_understanding/prediction_explanations/_algorithms.py:118]: Do we have test coverage for this logic?

**Review by dsherry — APPROVED** [review]
@chukarsten looking good to me :) left a couple questions

File: evalml/model_understanding/force_plots.py
Comment (bchen1116) [inline evalml/model_understanding/force_plots.py:86]: Nice! I like this

File: evalml/tests/model_understanding_tests/prediction_explanations_tests/test_force_plots.py
Comment (bchen1116) [inline evalml/tests/model_understanding_tests/prediction_explanations_tests/test_force_plots.py:70]: What's the difference between this and `initjs.assert_called()`?

File: evalml/tests/model_understanding_tests/prediction_explanations_tests/test_force_plots.py
Comment (bchen1116) [inline evalml/tests/model_understanding_tests/prediction_explanations_tests/test_force_plots.py:196]: why do we assert `len(class_labels)==1` here? Does this imply that there's only 1 class we're doing this for on regression pipelines?

**Review by bchen1116 — COMMENTED** [review]
Looking good! Left a few questions and clean-up comments.

**Pull request merged**
